### DISCUSSION TITLE
GeoServer Plugin

### DISCUSSION THREAD

**Discussion by weskamm** [geostyler/geostyler#1324] [General]
We need to get the GeoServer Plugin running better and become more visible at all.
Therefore i did a quick test of the current state.

My current findings.
Tested with latest stable 2.18.2:

Issues:

  * open geostyler tab on a rastersymbolizer SLD
    * geostyler does not appear
    * layerpreview does not appear
    * fontsize seems to be changed on the whole page -> bad css?!
    * console throws warns and erros:
      * Warn: @babel/polyfill is loaded more than once
      * Error: Wicket.Ajax: Cannot bind a listener for event "click" on element "idfa" because the element is not in the DOM
    * WFS Data Parser does a describefeature call for a raster style -> leads to errors 
https://github.com/geoserver/geoserver/blob/051d7a91c3dae51e18b71a344f3867687c722b2b/src/community/geostyler/src/main/resources/org/geoserver/wms/web/data/geostyler-init.ftl#L19
    * Error: Uncaught (in promise) Could not parse XML document: TypeError: Cannot read property 'forEach' of undefined

  * open geostyler tab on a polygonsymbolizer SLD
    * geostyler appears
    * layerpreview does not appear
    * fontsize seems to be changed on the whole page -> bad css?!
    * console throws warns and erros:
      * Warn: @babel/polyfill is loaded more than once
    * Some Workflow logs to console "onchange" and the SLD body
        see https://github.com/geoserver/geoserver/blob/051d7a91c3dae51e18b71a344f3867687c722b2b/src/community/geostyler/src/main/resources/org/geoserver/wms/web/data/geostyler-init.ftl#L58
    * CSS seems to be bad when entering fullscreen mode, not sure if this related to the plugin or a generic issue (can be also seen on default style tab)

Code needs overall cleanup, linting and update

Ideas:

  * Make SLD Editor collapsible? -> Size (height) can become an issue with geostyler and SLD editor shown at the same time

**Comment by weskamm** [comment]
I will start to fixup the issues mentioned.
Please discuss options / features for better integration and user experience or mention other issues you have found

**Comment by jansule** [comment]
Additionally, we should get rid of the `GeoStyler` tab and integrate GeoStyler properly (always above the SLD editor).

Another BIG thing is the proper support for the different SLD versions, as mentioned in https://github.com/geostyler/geostyler-sld-parser/issues/288

**Comment by weskamm** [comment]
I like the idea of integrating it directly in the default style tab, will have a look at that

**Reply by dnlkoch** [reply]
Yes, that would be awesome!

**Reply by weskamm** [reply]
Solution should look something like this, but i am still finding a way solving this in a clean way

![image](https://user-images.githubusercontent.com/1381363/107537806-948d1200-6bc3-11eb-8315-8bdfe44ddc5d.png)

**Reply by weskamm** [reply]
After discussing this on the dev mailinglist i think i can proceed with that idea.
Current state of work can be found here: https://github.com/geoserver/geoserver/compare/master...weskamm:geostylerintegration

**Comment by weskamm** [comment]
I think an option for "live updating" is a nice idea too!  

![geostyler](https://user-images.githubusercontent.com/1381363/108238396-358e4680-7149-11eb-89a5-197fd6344783.gif)

**Comment by weskamm** [comment]
Ok, issues have been fixed and current state can be found at
https://github.com/geoserver/geoserver/compare/master...weskamm:geostylerintegration

Needs following PR in geoserver before:
https://github.com/geoserver/geoserver/pull/4779

This is what it looks like:
![image](https://user-images.githubusercontent.com/1381363/109125646-94a51a00-774c-11eb-8be7-785f764d7a4d.png)
![image](https://user-images.githubusercontent.com/1381363/109125840-d766f200-774c-11eb-8e15-d64f4e33ee7c.png)

**Reply by KaiVolland** [reply]
Looks awesome. Can you switch between the side-by-side view and the below-and-above view?

**Reply by weskamm** [reply]
yes, side by side is the fullscreen view. i cropped the screenshot no. 1, it contains the menu as usual

**Comment by weskamm** [comment]
Final PR is out here: https://github.com/geoserver/geoserver/pull/4795
